test/common/upstream/bounded_load_hlb_test.cc

```diff
@@ -0,0 +1,250 @@
+  // In this test hosts 0, 1 & 2 are overloaded. The random shuffle
+  // sequence of 5 elements with seed 2 is 2 0 3 1 4. When the host
+  // picked up for hash 2 (which is 127.0.0.12) is overloaded, the
+  // method passes over host 0 and picks host 3 (127.0.0.13) up.
+
+  // setup: 5 hosts, few of them are overloaded.
+  std::vector<std::string> addresses;
+  addresses.push_back("127.0.0.10:90");
+  addresses.push_back("127.0.0.11:90");
+  addresses.push_back("127.0.0.12:90");
+  hostOverLoadedPredicate_ = getHostOverloadedPredicate(addresses);
+
+  NormalizedHostWeightVector normalized_host_weights;
+  createHosts(5, normalized_host_weights);
+
+  NormalizedHostWeightVectorConstPtr ring =
+      std::make_unique<const NormalizedHostWeightVector>(normalized_host_weights);
+  hlb_ = std::make_shared<TestHashingLoadBalancer>(ring);
+
+  NormalizedHostWeightVectorConstPtr normalized_host_weights_ptr =
+      std::make_unique<const NormalizedHostWeightVector>(normalized_host_weights);
+  lb_ = std::make_unique<TestBoundedLoadHashingLoadBalancer>(hlb_, normalized_host_weights_ptr, 1,
+                                                             hostOverLoadedPredicate_);
+
+  // test
+  HostConstSharedPtr host = lb_->chooseHost(2, 1);
+  EXPECT_NE(host, nullptr);
+  EXPECT_EQ(host->address()->asString(), "127.0.0.13:90");
+};
+
+// Works correctly for the case when requests with different hash map to the same
+// overloaded host.
+TEST_F(BoundedLoadHashingLoadBalancerTest, MultipleHashSameHostOverloaded) {
+  // In this case host 3 is overloaded and the CH ring has same host repeated on
+  // consecutive indices (0 0 1 1 2 2 3 3 4 4). The hashes 6 and 7 map to same host
+  // 3 which is overloaded. The random shuffle sequence of 5 elements with seed 6 is
+  // 4 0 1 2 3 and with 7 it is 0 1 2 3 4. Hence hosts 4 and 0 are picked up for these
+  // hashes.
+
+  // setup: 5 hosts, one of them is overloaded.
+  std::vector<std::string> addresses;
+  addresses.push_back("127.0.0.13:90");
+  hostOverLoadedPredicate_ = getHostOverloadedPredicate(addresses);
+
+  NormalizedHostWeightVector normalized_host_weights, hosts_on_ring;
+  createHostsMappedByMultipleHosts(5, normalized_host_weights, hosts_on_ring);
+
+  NormalizedHostWeightVectorConstPtr ring =
+      std::make_unique<const NormalizedHostWeightVector>(hosts_on_ring);
+  hlb_ = std::make_shared<TestHashingLoadBalancer>(ring);
+
+  NormalizedHostWeightVectorConstPtr normalized_host_weights_ptr =
+      std::make_unique<const NormalizedHostWeightVector>(normalized_host_weights);
+  lb_ = std::make_unique<TestBoundedLoadHashingLoadBalancer>(hlb_, normalized_host_weights_ptr, 1,
+                                                             hostOverLoadedPredicate_);
+
+  // test
+  HostConstSharedPtr host1 = lb_->chooseHost(6, 1);
+  EXPECT_NE(host1, nullptr);
+  HostConstSharedPtr host2 = lb_->chooseHost(7, 1);
+  EXPECT_NE(host2, nullptr);
+
+  // they are different
```

Comment:
comments here arent very useful, either obvious (EXEPCT_NE means we're testing that they are different) or unclear (how does the sequence value relate to the actual assertion?)